subprojects/core/src/main/java/org/gradle/internal/progress/DefaultBuildOperationExecutor.java

```diff
@@ -124,6 +128,8 @@ public void run(RunnableBuildOperation buildOperation) {
     }
 
     private <O extends BuildOperation> void executeInParallel(BuildOperationQueue.QueueWorker<O> worker, Action<BuildOperationQueue<O>> queueAction) {
+        failIfInResourceLockTransform();
```

Comment:
I think we should prevent any work, not just parallel. It would complete but it's still a performance pitfall to do expensive stuff under that lock.